ytypes/node.go

```diff
@@ -341,6 +341,27 @@ func retrieveNodeList(schema *yang.Entry, root interface{}, path, traversedPath
 
 	listKeyT := rv.Type().Key()
 	listElemT := rv.Type().Elem()
+
+	listKeyFieldNameMap := map[int]string{}
+	schemaKeyMap := map[string]string{}
+
+	// Reduce FieldByName calls in the loop and reduce Field calls and tag lookup calls in the loop.
+	if listKeyT.Kind() == reflect.Struct {
+		for i := 0; i < listKeyT.NumField(); i++ {
+			fieldName := listKeyT.Field(i).Name
+			if elem, ok := listElemT.Elem().FieldByName(fieldName); ok {
+				schemaKey, err := directDescendantSchema(elem)
+				if err != nil {
+					return nil, status.Errorf(codes.Unknown, "unable to get direct descendant schema name for %v: %v", schemaKey, err)
+				}
+
+				schemaKeyMap[fieldName] = schemaKey
+			}
```

Comment:
To keep the error statement close to where it was generated, you can return an error when ok=False like the original code
```go
if !ok {
				return nil, status.Errorf(codes.NotFound, "element struct type %v does not contain key field %s", listElemT, fieldName)
			}
```